File: temple/src/datastore/project.rs
use crate::{
    datastore::Repo,
    model::{Project, Uuid},
    util, AppResult,
};

pub async fn list(repo: &Repo) -> AppResult<Vec<Project>> {
    let projects = repo.select_projects().await?;

    Ok(projects)
}

pub struct CreateProjectAttributes {
    pub name: String,
    pub description: String,
}

pub async fn create(repo: &Repo, attributes: CreateProjectAttributes) -> AppResult<Project> {
    let CreateProjectAttributes { name, description } = attributes;

    let project = repo
        .create_project(Project {
            id: Default::default(),
            create_time: Default::default(),
            slug: util::slug::sluggify(&name),
            name,
            description,
        })
        .await?;

    Ok(project)
}

pub async fn get(repo: &Repo, id: Uuid) -> AppResult<Project> {
    let project = repo.get_project(id).await?;

    Ok(project)
}

pub async fn find(repo: &Repo, slug: &str) -> AppResult<Project> {
    let project = repo.find_project(slug).await?;

    Ok(project)
}

pub async fn update(repo: &Repo, project: Project) -> AppResult<Project> {
    let project = repo.update_project(project).await?;

    Ok(project)
}

File: theater/app/models/project.rb
class Project
  include ActiveModel::API
  include ActiveModel::Validations

  attr_accessor :description, :name, :slug

  validates :name, presence: true

  def id
    @slug
  end

  def persisted?
    id.present?
  end

  def save
    return if invalid?

    proto_project = ProjectsApi.new.create_project(self).project
    @slug = proto_project.slug
    Project.from_proto(proto_project)
  end

  def rename(new_name)
    @name = new_name

    return if invalid?

    proto_project = ProjectsApi.new.rename_project(self).project
    @slug = proto_project.slug
    Project.from_proto(proto_project)
  end

  def archive
    ProjectsApi
      .new
      .archive_project(self)
  end

  def restore
    ProjectsApi
      .new
      .restore_project(self)
  end

  def destroy
    ProjectsApi
      .new
      .delete_project(self)
  end

  def models
    ProjectsApi
      .new
      .list_models(self)
      .models
      .map { |proto_model| Model.from_proto(proto_model) }
  end

  # rubocop:todo Metrics/MethodLength
  def find_model(model_id)
    response = ProjectsApi.new.get_model(self, model_id)

    model = Model.from_proto(response.model)
    model.attributes = response.attributes.map { |proto_attribute| ModelAttribute.from_proto(proto_attribute) }
    model.associations = response.associations.map do |proto_association|
      ModelAssociation.from_proto(proto_association)
    end

    model
  end
  # rubocop:enable Metrics/MethodLength

  class << self
    def from_proto(proto_project)
      Project.new(
        description: proto_project.description,
        name: proto_project.name,
        slug: proto_project.slug,
      )
    end

    def all
      ProjectsApi
        .new
        .list_projects
        .projects
        .map { |proto_project| Project.from_proto(proto_project) }
    end

    def archived
      ProjectsApi
        .new
        .list_archived_projects
        .projects
        .map { |proto_project| Project.from_proto(proto_project) }
    end

    def find(id)
      proto_project = ProjectsApi.new.get_project(id).project

      Project.from_proto(proto_project)
    end
  end
end
